sermant-plugins/sermant-mq-grayscale/sermant-mq-grayscale-plugin/src/main/java/io/sermant/mq/grayscale/utils/SubscriptionDataUtils.java

```diff
@@ -0,0 +1,418 @@
+    public static void resetAutoCheckGrayTagItems(List<GrayTagItem> grayTagItems, MqConsumerClientConfig clientConfig) {
+        String addrTopicGroupKey = buildAddrTopicGroupKey(clientConfig.getTopic(), clientConfig.getConsumerGroup(),
+                clientConfig.getAddress());
+        AUTO_CHECK_GRAY_TAGS.remove(addrTopicGroupKey);
+        setAutoCheckTagChangeMap(clientConfig.getAddress(), clientConfig.getTopic(), clientConfig.getConsumerGroup(),
+                true);
+        if (!grayTagItems.isEmpty()) {
+            AUTO_CHECK_GRAY_TAGS.put(addrTopicGroupKey, grayTagItems);
+        }
+    }
+
+    /**
+     * set base consumer address@topic@group correspondents change flag
+     *
+     * @param namesrvAddr namesrvAddr
+     * @param topic topic
+     * @param group group
+     * @param flag flag
+     */
+    public static void setAutoCheckTagChangeMap(String namesrvAddr, String topic, String group, boolean flag) {
+        String addrTopicGroupKey = buildAddrTopicGroupKey(topic, group, namesrvAddr);
+        BASE_SUBSCRIPT_GRAY_TAG_CHANGE_MAP.put(addrTopicGroupKey, flag);
+    }
+
+    /**
+     * set gray consumer address@topic@group correspondents change flag
+     *
+     * @param namesrvAddr namesrvAddr
+     * @param topic topic
+     * @param group group
+     * @param flag flag
+     */
+    public static void setGrayGroupTagChangeMap(String namesrvAddr, String topic, String group, boolean flag) {
+        String addrTopicGroupKey = buildAddrTopicGroupKey(topic, group, namesrvAddr);
+        GRAY_GROUP_TAG_CHANGE_MAP.put(addrTopicGroupKey, flag);
+    }
+
+    /**
+     * using namesrvAddr/topic/consumerGroup build key
+     *
+     * @param topic topic
+     * @param consumerGroup consumerGroup
+     * @param namesrvAddr namesrvAddr
+     * @return namesrvAddr@topic@consumerGroup
+     */
+    public static String buildAddrTopicGroupKey(String topic, String consumerGroup, String namesrvAddr) {
+        String topicTemp = topic.contains(RETYPE) ? StringUtils.substringAfterLast(topic, RETYPE) : topic;
+        String consumerGroupTemp = consumerGroup.contains(RETYPE)
+                ? StringUtils.substringAfterLast(consumerGroup, RETYPE) : consumerGroup;
+        return namesrvAddr + AFA_SYMBOL + topicTemp + AFA_SYMBOL + consumerGroupTemp;
+    }
+
+    /**
+     * get gray tag change flag
+     *
+     * @param topic topic
+     * @param rebalance rebalance
+     * @return changeFlag
+     */
+    public static boolean getGrayTagChangeFlag(String topic, RebalanceImpl rebalance) {
+        String addrTopicGroupKey = buildAddrTopicGroupKey(topic, rebalance.getConsumerGroup(),
+                rebalance.getmQClientFactory().getClientConfig().getNamesrvAddr());
+        if (StringUtils.isEmpty(MqGrayscaleConfigUtils.getGrayGroupTag())) {
+            return BASE_SUBSCRIPT_GRAY_TAG_CHANGE_MAP.get(addrTopicGroupKey) != null
+                    && BASE_SUBSCRIPT_GRAY_TAG_CHANGE_MAP.get(addrTopicGroupKey);
+        } else {
```

Comment:
The above has already been returned. There is no need to add else. Just write the logic in else directly.

Reply:
fixed